Cse p4 enablement (#4079)
sourcegraph/about: handbook/support/p4-enablement.md

> # Dogfood Perforce Server

Reviewer: ```suggestion
# Dogfood Perforce server
```

> Perforce is a version control system like git, subversion, or mercurial. While git is based on a distributed, decentralised model, Perforce is centralised. For testing purposes you may use our perforce dogfood server.

Reviewer: ```suggestion
Perforce is a version control system like Git, subversion, or mercurial. While git is based on a distributed, decentralised model, Perforce is centralised. For testing purposes, you may use our Perforce dogfood server.
```

> ## Setting Up

Reviewer: ```suggestion
## Setting up
```

> - To connect to perforce you'll need the perforce cli installed locally. Use the command: `brew install --cask perforce`

Reviewer: ```suggestion
- To connect to the Perforce server, you'll need the Perforce cli installed locally. Use the command: `brew install --cask perforce`
```

> - The following environment variables configure your shell to point at the perforce server. Set them to your `env` with the `export` command, or add them to your `.bashrc` or `.zshrc` file.

Reviewer: ```suggestion
- The following environment variables configure your shell to point at the Perforce server. Set them to your `env` with the `export` command, or add them to your `.bashrc` or `.zshrc` file.
```

```
# .zshrc or .bashrc

> P4PORT=perforce.sgdev.org:1666 # This points the p4 cli at the dogfood server

Reviewer 2: Could we also add a link here to the actual VM that is running Perforce?

Author: Happy to add info about this

@unknwon Is the perforce server hosted on GCP? If so, can you let us know the project name?

Reviewer 3: See here for info about the deployment https://k8s.sgdev.org/github.com/sourcegraph/infrastructure/-/tree/dogfood/kubernetes/tooling/perforce

Author: Thanks Dave, adding links

> # .zshrc or .bashrc
> 
> P4PORT=perforce.sgdev.org:1666 # This points the p4 cli at the dogfood server
> P4USER=admin # sets your user
> P4PASSWD=<login session ticket value explained below> # doesn't require string quotes
> P4EDITOR=/usr/bin/vim # specifies the editor opened by some p4 commands

Reviewer: ```suggestion
# .zshrc or .bashrc

P4PORT=perforce.sgdev.org:1666 # This points the p4 cli at the dogfood server
P4USER=admin                   # sets your user
P4PASSWD=<session ticket>      # see details below, doesn't require string quotes
P4EDITOR=/usr/bin/vim          # specifies the editor opened by some p4 commands
```

```

> ## Interacting With Perforce Dogfood

Reviewer: ```suggestion
## Interacting with Perforce dogfood
```

To add repos to the perforce dogfood server follow the following procedure:
- [Dogfood Perforce Server](#dogfood-perforce-server)
  - [Setting Up](#setting-up)
  - [Interacting With Perforce Dogfood](#interacting-with-perforce-dogfood)
    - [Generate A Session Ticket](#generate-a-session-ticket)
    - [Create New Depot](#create-new-depot)
    - [Create A Client](#create-a-client)
    - [Add Files and Submit](#add-files-and-submit)
    - [Configuring Sourcegraph To Sync Dogfood Depots](#configuring-sourcegraph-to-sync-dogfood-depots)
- [Local Perforce Server](#local-perforce-server)

> ### Generate A Session Ticket

Reviewer: We use sentence case in headings (not going to suggest changes to remaining of this doc 😁 )

Author: Thanks for giving it a thorough look over 🙏

Interaction with the dogfood server requires authentication. Once you've set up your shell environment you'll need to generate a session ticket. To do this you'll need your users password. You can find the [admin password](https://team-sourcegraph.1password.com/vaults/dnrhbauihkhjs5ag6vszsme45a/allitems/fac6hoq3ujb3xpxtllbijzyxta), and others on our shared 1Password account.

> 1. Once your environment is set run `p4 ping` this will prompt you for the admin password, and is a good way to test your connection to the server.

Reviewer: If you have configured `P4PASSWD` correctly, `p4` will not ask for a password again and just take it. However, by default session ticket has a lifetime of 12 hours, so in the interactive mode (ie. terminal), `P4PASSWD` is actually not as useful/needed.

Author: 🤔  I'm not sure I understand this, my work flow involved setting the session ticket to the env variable P4PASSWD, which doesn't require updating for like 12 hours as you've noted. If I didn't most p4 commands would regularly require entering the admin password. Is there a more efficient way to do this I've overlooked? 

I think I can clarify whats already there a bit more though 👍

Reviewer: By default, every time you entered a password interactively, a session ticket is stored directly in your home directory `~/.p4tickets`, which is the same value as you would put as `P4PASSWD`.

But having an extra step does not hurt 😁

Reviewer: > If I didn't most p4 commands would regularly require entering the admin password.

I think the reason for this because you do have `P4PASSWD` set, either invalid or set with empty string.

Author: Ahh yeah you're right, I set P4PASSWD early, and only ever ran `p4 login` with the -p flag which prevents storage to `.p4tickets`

Author: added a note about alternative login with .p4tickets

Files:
        //support-tools/.git/COMMIT_EDITMSG     # add
        //support-tools/.git/FETCH_HEAD # add
        //support-tools/.git/HEAD       # add
        //support-tools/.git/ORIG_HEAD  # add
        //support-tools/.git/config     # add
        //support-tools/.git/description        # add

        # etc etc
```
You can see the files staged for adding with `p4 opened`.

Finally run `p4 submit` to load the files to the depo on the server.

### Configuring Sourcegraph to sync dogfood depots

Our [documentation](https://docs.sourcegraph.com/admin/repo/perforce) covers the requirements to sync to Sourcegraph, however for convienience it should be noted we have a user called `buildkite` on our dogfood instance who session ticket will not expire. To generate the ticket for this account reference our shared [1Password](https://team-sourcegraph.1password.com/vaults/dnrhbauihkhjs5ag6vszsme45a/allitems/lajspc6a5valfbsh3whpcb5bp4).

To learn more about general p4 commands checkout this resource:
[https://www.perforce.com/perforce/doc.973/cmdguide/html/quicksta.htm](https://www.perforce.com/perforce/doc.973/cmdguide/html/quicksta.htm)

# Local Perforce server

> Joe has also developed an awesome perforce server [image](https://github.com/sourcegraph/helix-docker/blob/main/helix-p4d/Dockerfile) that allows for local deployment of a perforce server.

Reviewer: ```suggestion
Joe has also developed an awesome Perforce server [image](https://github.com/sourcegraph/helix-docker/blob/main/helix-p4d/Dockerfile) that allows for local deployment of a Perforce server.
```

Please respect brand names 😁 Despite I do not like Perfoce so far.

Author: 😁

 
The following procedure runs the image:

1. Start the server `docker run --publish 1666:1666 sourcegraph/helix-p4d:2020.2`

2. Open a new terminal, configure the terminal session env:
```
export P4PORT=:1666
export P4USER=admin
```

> 3. ping the server from a separate terminal with p4 ping and enter the password below:

Reviewer: ```suggestion
3. Ping the server from a separate terminal with p4 ping and enter the password below:
```

```

> ARG P4PASSWD=pass12349ers

Reviewer: ```suggestion
pass12349ers
```

I am concerned it would confuse people that "ARG P4PASSWD=" is also part of the password.
